code_samples/catalog/custom_code_generator_strategy/src/CodeGenerator/Strategy/ProductTypeCodeGenerator.php

```diff
@@ -0,0 +1,23 @@
+<?php declare(strict_types=1);
+
+namespace App\CodeGenerator\Strategy;
+
+use Ibexa\Contracts\Core\Exception\InvalidArgumentException;
+use Ibexa\Contracts\ProductCatalog\Local\CodeGenerator\CodeGeneratorContext;
+use Ibexa\Contracts\ProductCatalog\Local\CodeGenerator\CodeGeneratorInterface;
+
+class ProductTypeCodeGenerator implements CodeGeneratorInterface
+{
+    public function generateCode(CodeGeneratorContext $context): string
+    {
+        if (!$context->hasBaseProduct()) {
+            throw new InvalidArgumentException('$context', 'missing base product');
+        }
+
+        if (!$context->hasIndex()) {
+            throw new InvalidArgumentException('$context', 'missing index');
+        }
+
+        return $context->getBaseProduct()->getProductType()->getIdentifier() . '-' . $context->getIndex();
```

Comment:
There is only one issue in this implementation: index is local to base product so codes are unique only for the first time. 

Due to lack of better alternative let's use the following example :

```
<?php

declare(strict_types=1);

namespace App\ProductCatalog\CodeGenerator;

use Ibexa\Contracts\Core\Exception\InvalidArgumentException;
use Ibexa\Contracts\ProductCatalog\Local\CodeGenerator\CodeGeneratorContext;
use Ibexa\Contracts\ProductCatalog\Local\CodeGenerator\CodeGeneratorInterface;

/**
 * Generates codes in the following format: <base_code>v<index>
 *
 * For example: "LP542v1"
 */
final class CustomIncrementalCodeGenerator implements CodeGeneratorInterface
{
    public function generateCode(CodeGeneratorContext $context): string
    {
        if (!$context->hasBaseProduct()) {
            throw new InvalidArgumentException('$context', 'missing base product');
        }

        if (!$context->hasIndex()) {
            throw new InvalidArgumentException('$context', 'missing index');
        }

        return $context->getBaseProduct()->getCode() . 'v' . $context->getIndex();
    }
}
```